graphix/open_graph.py

```diff
@@ -0,0 +1,383 @@
+        in_verts = add_vertices(len(self.inputs), zx.VertexType.BOUNDARY)
+        g.set_inputs(in_verts)
+
+        # Add nodes for internal Z spiders - not including the phase gadgets
+        body_verts = add_vertices(len(self._inside), zx.VertexType.Z)
+
+        # Add nodes for the phase gadgets. In OpenGraph we don't store the
+        # effect as a seperate node, it is instead just stored in the
+        # "measurement" attribute of the node it measures.
+        x_meas = [i for i, m in self.measurements.items() if m.plane == "YZ"]
+        x_meas_verts = add_vertices(len(x_meas), zx.VertexType.Z)
+
+        out_verts = add_vertices(len(self.outputs), zx.VertexType.BOUNDARY)
+        g.set_outputs(out_verts)
+
+        # Maps a node's ID in the Open Graph to it's corresponding node ID in
+        # the PyZX graph and vice versa.
+        map_to_og = dict(zip(body_verts, self._inside.nodes()))
+        map_to_pyzx = {v: i for i, v in map_to_og.items()}
+
+        # Open Graph's don't have boundary nodes, so we need to connect the
+        # input and output Z spiders to their corresponding boundary nodes in
+        # pyzx.
+        for pyzx_index, og_index in zip(in_verts, self.inputs):
+            g.add_edge((pyzx_index, map_to_pyzx[og_index]))
+        for pyzx_index, og_index in zip(out_verts, self.outputs):
+            g.add_edge((pyzx_index, map_to_pyzx[og_index]))
+
+        og_edges = self._inside.edges()
+        pyzx_edges = [(map_to_pyzx[a], map_to_pyzx[b]) for a, b in og_edges]
+        g.add_edges(pyzx_edges, zx.EdgeType.HADAMARD)
+
+        # Add the edges between the Z spiders in the graph body
+        for og_index, meas in self.measurements.items():
+            # If it's an X measured node, then we handle it in the next loop
+            if meas.plane == "XY":
+                g.set_phase(map_to_pyzx[og_index], meas.angle)
+
+        # Connect the X measured vertices
+        for og_index, pyzx_index in zip(x_meas, x_meas_verts):
+            g.add_edge((map_to_pyzx[og_index], pyzx_index), zx.EdgeType.HADAMARD)
+            g.set_phase(pyzx_index, self.measurements[og_index].angle)
+
+        return g
+
+    @classmethod
+    def from_pyzx_graph(cls, g: zx.graph.base.BaseGraph) -> OpenGraph:
+        """Constructs an Optyx Open Graph from a PyZX graph.
+
+        This method may add additional nodes to the graph so that it adheres
+        with the definition of an OpenGraph. For instance, if the final node on
+        a qubit is measured, it will add two nodes behind it so that no output
+        nodes are measured to satisfy the requirements of an open graph.
+
+        Example
+        -------
+        >>> import pyzx as zx
+        >>> from graphix.open_graph import OpenGraph
+        >>> circ = zx.qasm("qreg q[2]; h q[1]; cx q[0], q[1]; h q[1];")
+        >>> g = circ.to_graph()
+        >>> og = OpenGraph.from_pyzx_graph(g)
+        """
+        zx.simplify.to_graph_like(g)
+        zx.simplify.full_reduce(g)
```

Comment:
don't we want to check that `g` is graph_like, rather than performing `full_reduce` automatically? I think it's up to user what type of graph they want to translate (e.g. may not always want to full-reduce)

Reply:
Yes, sorry I think this is a remnant of the benchmarking that we ran before. I will remove it